proxy/gateway_test.go

```diff
@@ -0,0 +1,224 @@
+				"data": "InJlc3AtZGF0YSI="
+			}]
+		}
+	}`, string(httpRespBody))
+
+	// Now try with shorthand payload notation
+	srv.lastQueryReq = nil
+	httpResp, err = http.Post(
+		fmt.Sprintf(
+			"http://%v/api/v1/namespaces/my-namespace/workflows/my-workflow/query/my-query",
+			srv.httpListener.Addr(),
+		),
+		"application/json",
+		strings.NewReader(`{ "query": { "queryArgs": ["req-data"] } }`),
+	)
+	require.NoError(t, err)
+	httpRespBody, err = io.ReadAll(httpResp.Body)
+	httpResp.Body.Close()
+	require.NoError(t, err)
+	require.Equal(t, httpResp.StatusCode, http.StatusOK, "Bad status: %s", httpRespBody)
+	require.True(t, proto.Equal(srv.lastQueryReq, grpcReq))
+	require.JSONEq(t, `{ "queryResult": ["resp-data"] }`, string(httpRespBody))
+}
+
+type testServer struct {
+	workflowservice.UnimplementedWorkflowServiceServer
+	grpcListener net.Listener
+	grpcServer   *grpc.Server
+	httpListener net.Listener
+	httpServer   http.Server
+	serveErrCh   chan error
+
+	lastQueryReq  *workflowservice.QueryWorkflowRequest
+	nextQueryResp *workflowservice.QueryWorkflowResponse
+	nextQueryErr  error
+}
+
+var shorthandMarshaler runtime.Marshaler
+var nonShorthandMarshaler runtime.Marshaler
+
+func init() {
+	if m, err := proxy.NewJSONPBMarshaler(proxy.JSONPBMarshalerOptions{}); err != nil {
+		panic(err)
+	} else if u, err := proxy.NewJSONPBUnmarshaler(proxy.JSONPBUnmarshalerOptions{}); err != nil {
+		panic(err)
+	} else {
+		shorthandMarshaler = proxy.NewGRPCGatewayJSONPBMarshaler(m, u)
+	}
+	if m, err := proxy.NewJSONPBMarshaler(proxy.JSONPBMarshalerOptions{DisablePayloadShorthand: true}); err != nil {
+		panic(err)
+	} else if u, err := proxy.NewJSONPBUnmarshaler(proxy.JSONPBUnmarshalerOptions{DisablePayloadShorthand: true}); err != nil {
+		panic(err)
+	} else {
+		nonShorthandMarshaler = proxy.NewGRPCGatewayJSONPBMarshaler(m, u)
+	}
+}
+
+func StartTestServer(ctx context.Context) (*testServer, error) {
+	serveMux := runtime.NewServeMux(
+		runtime.WithMarshalerOption("application/json+no-payload-shorthand", nonShorthandMarshaler),
+		runtime.WithMarshalerOption(runtime.MIMEWildcard, shorthandMarshaler),
+	)
+	t := &testServer{
+		serveErrCh: make(chan error, 2),
+		grpcServer: grpc.NewServer(),
+	}
+	t.httpServer.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if _, ok := r.URL.Query()["noPayloadShorthand"]; ok {
+			r.Header.Set("Accept", "application/json+no-payload-shorthand")
+		}
+		serveMux.ServeHTTP(w, r)
+	})
+
+	// Start gRPC server
+	var err error
+	if t.grpcListener, err = net.Listen("tcp", "127.0.0.1:0"); err != nil {
+		return nil, err
+	}
+	workflowservice.RegisterWorkflowServiceServer(t.grpcServer, t)
+	go func() { t.serveErrCh <- t.grpcServer.Serve(t.grpcListener) }()
+
+	// Start HTTP server
+	if t.httpListener, err = net.Listen("tcp", "127.0.0.1:0"); err != nil {
+		t.grpcServer.Stop()
+		return nil, err
+	}
+	if err = workflowservice.RegisterWorkflowServiceHandlerServer(ctx, serveMux, t); err != nil {
```

Comment:
I will push the generated gateway code when https://github.com/temporalio/api/pull/301 is reviewed